### ISSUE TITLE
alt-related key-combination seems not working in macOS.

### ISSUE THREAD

**Issue by hzb-ykpcx** [helix-editor/helix#2469]
I tried to use option key combined with other keys to use the command such as alt ` and alt-, but it didn't work. I am using MacBook Pro 13.3 M1.

**Comment by the-mikedavis** [comment]
Are you using iTerm2? You may need to configure it so it doesn't consume the alt keypresses itself: https://www.clairecodes.com/blog/2018-10-15-making-the-alt-key-work-in-iterm2/

**Comment by hzb-ykpcx** [comment]
> Are you using iTerm2? You may need to configure it so it doesn't consume the alt keypresses itself: https://www.clairecodes.com/blog/2018-10-15-making-the-alt-key-work-in-iterm2/

actually I am using warp

**Comment by the-mikedavis** [comment]
What does `showkey -a` say when you press keys with alt held?

```
$ showkey -a
Press any keys - Ctrl-D will terminate this program

^[j 	27 0033 0x1b
 	106 0152 0x6a
^[k 	27 0033 0x1b
 	107 0153 0x6b
^[h 	27 0033 0x1b
 	104 0150 0x68
^[l 	27 0033 0x1b
 	108 0154 0x6c
^D 	 4 0004 0x04
```

(above is `A-j`, `A-k`, `A-h`, and `A-l`)

**Comment by ChuChuNaKu** [comment]
I'm running into the same issue. 

Mac OS Monterey version 12.3.1
Helix 22.03
Kitty version 0.25.0

Here is the result after running showkey -a and press keys with the alt key held. 
A-j, A-k, A-h, and A-l

b<ALT-CTL-H=BS><ALT-CTL-F=ACK>
K<ALT-CTL-Z=SUB>
K<ALT-CTL-Y=EM>
B,

**Comment by hzb-ykpcx** [comment]
> What does `showkey -a` say when you press keys with alt held?
> 
> ```
> $ showkey -a
> Press any keys - Ctrl-D will terminate this program
> 
> ^[j 	27 0033 0x1b
>  	106 0152 0x6a
> ^[k 	27 0033 0x1b
>  	107 0153 0x6b
> ^[h 	27 0033 0x1b
>  	104 0150 0x68
> ^[l 	27 0033 0x1b
>  	108 0154 0x6c
> ^D 	 4 0004 0x04
> ```
> 
> (above is `A-j`, `A-k`, `A-h`, and `A-l`)

here is what I got
showkey -a
Type any key to see the sequence it sends.
Terminate with your shell interrupt <CTL-C=ETX> or quit <FS> character.
b<ALT-CTL-H=BS><ALT-CTL-F=ACK>K<ALT-CTL-Z=SUB>K<ALT-CTL-Y=EM>B,

**Comment by ChuChuNaKu** [comment]
I also tried adding [this to my Kitty config](https://github.com/kovidgoyal/kitty/issues/847#issuecomment-417478876) but that didn't seem to work.

**Comment by hzb-ykpcx** [comment]
> I solved my issue. The problem was with the Kitty terminal. I first tried adding [this to my Kitty config](https://github.com/kovidgoyal/kitty/issues/847#issuecomment-417478876) but that didn't seem to work. Then I tried setting the config to `macos_option_as_alt yes` and that fixed it for me.
> 
> [This _might_ help for Warp](https://github.com/warpdotdev/warp/issues/512#issuecomment-1014072003).

I set the left option key as meta while not changing the right option key, and strange things happened.  
when I use the left option key(meta), I cannot jump over words in the prompt with <alt-arrow> key , however **part of** the helix editor key combination with alt key works such as <alt-d>. But for <alt-`>, it still doesn't work.  
When I use the right option key (non meta), I can jump over words with option + arrow key, but None of the <alt-_key_> combination works. 
What I expect is that you can't use any key combination with meta key, but with unmodified alt(option) everything should work. I guess there's a bug?

**Comment by hzb-ykpcx** [comment]
I set my left option key as meta key. And I find most of the alt-key combinations work, except: `alt-~` `alt-u` `alt-U`.
And why does meta key act as alt?

**Comment by ChuChuNaKu** [comment]
I was going to suggest checking with the Warp Issue queue but [it looks like you beat me to it](https://github.com/warpdotdev/Warp/issues/1454). Just as a heads up, someone added an [entry to the Helix wiki ](https://github.com/helix-editor/helix/wiki/FAQ#alt-key-combinations-do-not-work-on-macos-iterm2) if you decide to try a different terminal.

**Comment by BGR360** [comment]
see also https://github.com/helix-editor/helix/issues/2280

**Comment by erasin** [comment]
@hzb-ykpcx  

> I set my left option key as meta key. And I find most of the alt-key combinations work, except: `alt-~` `alt-u` `alt-U`.


[macos help: Use dead keys](https://support.apple.com/en-hk/guide/mac-help/mh27474/mac) 

![mac deadkey](https://help.apple.com/assets/62A8ED2F99A5D0045E612A42/62A8ED3DCA32D1047032C8FF/en_GB/23a22194a2aafd3e59a086b1931916fc.png)

It can change the keyboard  layout by [ukelele](https://scripts.sil.org/ukelele) 

https://github.com/alacritty/alacritty/issues/458#issuecomment-562992719

**Comment by gbinux** [comment]
I had the same issue (using terminal.app). Checking the box "Use option as Meta key"  (see snapshot below) fixed my problem 

<img width="714" alt="image" src="https://user-images.githubusercontent.com/13037898/189510695-42f4e632-a8da-4352-81c1-69cadddb42ba.png">

**Comment by hzb-ykpcx** [comment]
Hey guys, thanks for your help. I think I should clarify the problem existing now. For iterm2, we can just set option key as "esc+" in the preferences as the picture shows, and no issue occurs. But for Warp, there are still some of the combinations that cannot work, like _A-` A-u A-U`_ , and I don't know why. I have posted an issue on warp repos and get no reply for now.
<img width="1084" alt="Screenshot 2022-09-11 at 22 59 17" src="https://user-images.githubusercontent.com/52552971/189534698-9829dd0a-4112-4cc2-9c4b-687d512c548b.png">

**Comment by matthiasdebernardini** [comment]
@hzb-ykpcx I tried your suggesting and it worked for me.

I'm on Monterey with a M1 using iTerm2

**Comment by theteachr** [comment]
Not sure of all the problems, but I'm able to use <kbd>⌘</kbd> <kbd>;</kbd> (Cmd + ;) to alternate the cursor between the start and end of the selection, for which Helix says that it's <kbd>Alt</kbd> <kbd>;</kbd>.

Although, https://github.com/helix-editor/helix/issues/2469#issuecomment-1127062149 works well (for Kitty).

**Comment by grafoo** [comment]
At least when using Alacritty, the config is quite straight forward.

E.g. to make the default keybinding for `switch_to_uppercase` work you'd use
```
key_bindings:
  - { key: Grave, mods: Alt, chars: "\x1b`" }
```
in the Alacritty config.

See https://github.com/alacritty/alacritty/issues/93#issuecomment-1364783147 for a more or less complete listing.

**Comment by victororlyk** [comment]
same for me with Alacritty

**Comment by joshbainbridge** [comment]
I think this is relevant. A six year old issue in Alacritty has just been resolved with an upstream change to winit:

https://github.com/alacritty/alacritty/commit/d9ba129e9e722a551cc0cc4c1899c97593ca26b4

This allows the Option key to behave as Alt. An example `alacritty.yml` will look like:

```yml
window:
  # Make `Option` key behave as `Alt` (macOS only):
  #   - OnlyLeft
  #   - OnlyRight
  #   - Both
  #   - None (default)
  option_as_alt: OnlyLeft
```

Not released yet, but I'd imagine this will be out in the next version. These are the related issues leading to the change:

https://github.com/alacritty/alacritty/issues/62
https://github.com/rust-windowing/winit/pull/2576

**Comment by melMass** [comment]
Replacing alt with cmd has a few issues:
- You cannot use tilde (~) anymore
- On an AZERTY keyboard you cannot use brackets anymore

**Comment by ipochi** [comment]
Facing this issue as well. 

I've set the keymappings as suggested in the clairecodes website @the-mikedavis 
I'm using iTerm2

In showkey -a , when pressed with Option Key + h or j or k or l 

I get <ESC>(h | j | k | l )

**Comment by OscartGiles** [comment]
For anyone wondering how to fix in warp you can bring up the command pallet with CMD-p, and then search `Enable Left Option Key is Meta`. Set that and Alt should work in Helix.

**Comment by qballer** [comment]
for iterm2 users

go to prefences, profiles, keys, and change these definitions from normal to inclue escaping like in the image. 
![image](https://github.com/helix-editor/helix/assets/1002622/bf3bc444-2d91-4018-88dc-b142a2e18b0e)

**Comment by anka-213** [comment]
Would it be possible to remap the keys directly in helix instead of in the terminal? The option `Use Option as Meta key` in Terminal.app isn't really an option for me since I am using a Swedish keyboard layout, which means that I would lose access to the symbols `@$|[]~` which all require using the alt-key to write.

The optimal solution for me would be if helix just interpreted the symbol that is generated as the corresponding alt-something key in normal-mode while still giving the generated symbols in insert-mode. So for example, in normal-mode Ç would be interpreted as `Alt-C`, while in insert-mode it would insert a Ç symbol. 

I'm assuming that this would depend on your specific layout though, so either helix would have to detect the layout somehow or it would be a configuration option.

**Comment by pascalkuthe** [comment]
yeah lets close this in in favor of https://github.com/helix-editor/helix/issues/2280, that has a lot more details. I will link this back in that issue.
